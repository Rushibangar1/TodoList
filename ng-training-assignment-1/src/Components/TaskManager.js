import React, { useState } from 'react';
import TaskForm from './TaskForm/TaskForm';
import TaskList from './TaskList/TaskList';

const TaskManager = () => {
  const [tasks, setTasks] = useState([]);
  const [editingTask, setEditingTask] = useState(null);
  const [isFormVisible, setIsFormVisible] = useState(false);

  const handleTaskSubmit = (task) => {
    if (editingTask) {
      setTasks(tasks.map((t) => (t.id === editingTask.id ? { ...task, id: editingTask.id } : t)));
      setEditingTask(null);
    } else {
      const newTask = { ...task, id: new Date().getTime() };
      setTasks([...tasks, newTask]);
    }
    setIsFormVisible(false);
  };

  const handleEdit = (task) => {
    setEditingTask(task);
    setIsFormVisible(true);
  };

  const handleDelete = (taskId) => {
    const taskToDelete = tasks.find((task) => task.id === taskId);
    // Show confirmation popup
    if (window.confirm(`Are you sure you want to delete the task: "${taskToDelete.assignedTo}"?`)) {
      setTasks(tasks.filter((task) => task.id !== taskId));
    }
  };

  const handleCancel = () => {
    setEditingTask(null);
    setIsFormVisible(false);
  };

  const toggleForm = () => {
    setIsFormVisible(!isFormVisible);
  };

  return (
    <div className="slds-box slds-p-around_medium">
      <button 
        className="slds-button slds-button_brand slds-m-bottom_medium" 
        onClick={toggleForm}
      >
        {isFormVisible ? 'Cancel' : 'Add Task'}
      </button>
      {isFormVisible && (
        <TaskForm taskToEdit={editingTask} onSubmit={handleTaskSubmit} onCancel={handleCancel} />
      )}
      <TaskList tasks={tasks} onEdit={handleEdit} onDelete={handleDelete} />
    </div>
  );
};

export default TaskManager;
