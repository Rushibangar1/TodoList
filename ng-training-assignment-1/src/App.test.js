import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import TaskManager from './Components/TaskManager';


describe('TaskManager Component', () => {
  test('renders Add Task button', () => {
    render(<TaskManager />);
    const addButton = screen.getByText(/Add Task/i);
    expect(addButton).toBeInTheDocument();
  });

  test('shows TaskForm when Add Task button is clicked', () => {
    render(<TaskManager />);
    const addButton = screen.getByText(/Add Task/i);
    fireEvent.click(addButton);
    
    const taskForm = screen.getByLabelText(/Assigned To/i);
    expect(taskForm).toBeInTheDocument();
  });

  test('submits the form and adds a new task', () => {
    render(<TaskManager />);
    const addButton = screen.getByText(/Add Task/i);
    fireEvent.click(addButton);

    // Fill the form fields
    fireEvent.change(screen.getByLabelText(/Assigned To/i), { target: { value: 'User1' } });
    fireEvent.change(screen.getByLabelText(/Status/i), { target: { value: 'Pending' } });
    fireEvent.change(screen.getByLabelText(/Due Date/i), { target: { value: '2024-09-30' } });
    fireEvent.change(screen.getByLabelText(/Priority/i), { target: { value: 'High' } });
    fireEvent.change(screen.getByLabelText(/Comments/i), { target: { value: 'Test comment' } });
    
    // Submit the form
    fireEvent.click(screen.getByText(/Add Task/i));
    
    // Check if the task is added to the list
    expect(screen.getByText(/User1/i)).toBeInTheDocument();
    expect(screen.getByText(/Pending/i)).toBeInTheDocument();
    expect(screen.getByText(/2024-09-30/i)).toBeInTheDocument();
    expect(screen.getByText(/High/i)).toBeInTheDocument();
    expect(screen.getByText(/Test comment/i)).toBeInTheDocument();
  });

  test('edits an existing task', () => {
    render(<TaskManager />);
    const addButton = screen.getByText(/Add Task/i);
    fireEvent.click(addButton);

    // Fill the form fields for the first task
    fireEvent.change(screen.getByLabelText(/Assigned To/i), { target: { value: 'User1' } });
    fireEvent.change(screen.getByLabelText(/Status/i), { target: { value: 'Pending' } });
    fireEvent.change(screen.getByLabelText(/Due Date/i), { target: { value: '2024-09-30' } });
    fireEvent.change(screen.getByLabelText(/Priority/i), { target: { value: 'High' } });
    fireEvent.change(screen.getByLabelText(/Comments/i), { target: { value: 'Test comment' } });
    
    // Submit the form
    fireEvent.click(screen.getByText(/Add Task/i));

    // Now edit the task
    const editButton = screen.getByText(/Edit/i);
    fireEvent.click(editButton);
    
    // Change the comments
    fireEvent.change(screen.getByLabelText(/Comments/i), { target: { value: 'Updated comment' } });
    const updateButtons = screen.getAllByText(/Update/i);
    fireEvent.click(updateButtons[0]);  // Click the first "Update" button

    // Check if the updated comment is displayed
    expect(screen.getByText(/Updated comment/i)).toBeInTheDocument();
  });

  test('deletes a task', () => {
    render(<TaskManager />);
    const addButton = screen.getByText(/Add Task/i);
    fireEvent.click(addButton);

    // Fill the form fields for the first task
    fireEvent.change(screen.getByLabelText(/Assigned To/i), { target: { value: 'User1' } });
    fireEvent.change(screen.getByLabelText(/Status/i), { target: { value: 'Pending' } });
    fireEvent.change(screen.getByLabelText(/Due Date/i), { target: { value: '2024-09-30' } });
    fireEvent.change(screen.getByLabelText(/Priority/i), { target: { value: 'High' } });
    fireEvent.change(screen.getByLabelText(/Comments/i), { target: { value: 'Test comment' } });
    
    // Submit the form
    fireEvent.click(screen.getByText(/Add Task/i));

    // Delete the task
    const deleteButton = screen.getByText(/Delete/i);
    fireEvent.click(deleteButton);

    // Check if the task is no longer in the document
    expect(screen.queryByText(/User1/i)).not.toBeInTheDocument();
  });

  test('shows no tasks message when no tasks are available', () => {
    render(<TaskManager />);
    
    const noTasksMessage = screen.getByText(/No tasks available/i);
    expect(noTasksMessage).toBeInTheDocument();
  });
});
