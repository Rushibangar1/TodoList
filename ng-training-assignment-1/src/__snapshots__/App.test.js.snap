// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <header
      class="App-header"
    />
    <div
      class="slds-box slds-p-around_medium"
    >
      <button
        class="slds-button slds-button_brand slds-m-bottom_medium"
      >
        Add Task
      </button>
      <div
        class="slds-box slds-p-around_medium"
      >
        <h1
          style="font-weight: bold; font-size: 24px; text-align: left; margin-bottom: 20px;"
        >
          Task Management Application
        </h1>
        <table
          class="slds-table slds-table_bordered slds-table_fixed-layout"
        >
          <thead>
            <tr>
              <th
                scope="col"
              >
                <div
                  class="slds-truncate"
                >
                  Assigned To
                </div>
              </th>
              <th
                scope="col"
              >
                <div
                  class="slds-truncate"
                >
                  Status
                </div>
              </th>
              <th
                scope="col"
              >
                <div
                  class="slds-truncate"
                >
                  Due Date
                </div>
              </th>
              <th
                scope="col"
              >
                <div
                  class="slds-truncate"
                >
                  Priority
                </div>
              </th>
              <th
                scope="col"
              >
                <div
                  class="slds-truncate"
                >
                  Comments
                </div>
              </th>
              <th
                scope="col"
              >
                <div
                  class="slds-truncate"
                >
                  Actions
                </div>
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td
                class="slds-text-align_center"
                colspan="6"
              >
                No tasks available
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
